test/e2e/scenarios/hubble/steps/validate-drop-metric.go

```diff
@@ -0,0 +1,120 @@
+	"log"
+	"net/http"
+	"reflect"
+
+	promclient "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+var ErrNoMetricFound = fmt.Errorf("no metric found")
+
+const (
+	destinationKey = "destination"
+	sourceKey      = "source"
+	protcolKey     = "protocol"
+	reason         = "reason"
+)
+
+type ValidateHubbleDropMetric struct {
+	PortForwardedHubblePort string // presumably port-forwarded to a cilium pod
+	Source                  string
+	Protocol                string
+	Reason                  string
+}
+
+func (v *ValidateHubbleDropMetric) Run() error {
+	promAddress := fmt.Sprintf("http://localhost:%s/metrics", v.PortForwardedHubblePort)
+	ctx := context.Background()
+	pctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	validMetric := map[string]string{
+		destinationKey: "",
+		sourceKey:      v.Source,
+		protcolKey:     v.Protocol,
+		reason:         v.Reason,
+	}
+
+	metrics := map[string]*promclient.MetricFamily{}
+	scrapeMetricsFn := func() error {
+		log.Printf("attempting scrape metrics on %s", promAddress)
+		var err error
+		metrics, err = getPrometheusDropMetrics(promAddress)
+		if err != nil {
+			return fmt.Errorf("failed to get prometheus metrics: %w", err)
+		}
+
+		return nil
+	}
+
+	err := defaultRetrier.Do(pctx, scrapeMetricsFn)
+	if err != nil {
+		return fmt.Errorf("could not start port forward within %ds: %w	", defaultTimeoutSeconds, err)
+	}
+
+	if !verifyLabelsPresent(metrics, validMetric) {
+		return fmt.Errorf("failed to find metric matching %+v: %w", validMetric, ErrNoMetricFound)
+	}
+
+	log.Printf("found metric matching %+v\n", validMetric)
+	return nil
+}
+
+func verifyLabelsPresent(data map[string]*promclient.MetricFamily, validMetric map[string]string) bool {
+	for _, metric := range data {
+		if metric.GetName() == "hubble_drop_total" {
+			for _, metric := range metric.GetMetric() {
+
+				// get all labels and values on the metric
+				metricLabels := map[string]string{}
+				for _, label := range metric.GetLabel() {
+					metricLabels[label.GetName()] = label.GetValue()
+				}
+				if reflect.DeepEqual(metricLabels, validMetric) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func getPrometheusDropMetrics(url string) (map[string]*promclient.MetricFamily, error) {
+	client := http.Client{}
+	resp, err := client.Get(url) //nolint
+	if err != nil {
+		return nil, fmt.Errorf("HTTP request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP request failed with status: %v", resp.Status) //nolint:goerr113,gocritic
+	}
+
+	metrics, err := parseReaderPrometheusMetrics(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return metrics, nil
+}
+
+func parseReaderPrometheusMetrics(input io.Reader) (map[string]*promclient.MetricFamily, error) {
+	var parser expfmt.TextParser
+	return parser.TextToMetricFamilies(input) //nolint
+}
+
+func (v *ValidateHubbleDropMetric) Prevalidate() error {
```

Comment:
nit: the only way to differentiate between `Prevalidate()`'s is based on how it is instantiated. Is there any other way we can define these functions as they coexist within the same package.

Reply:
Also, what is the intent of the pre|post validate functions? Could this be a part of the [validate package](https://github.com/Azure/azure-container-networking/tree/master/test/validate)?

Reply:
The purpose of `Prevalidate()` is for quick tasks that can be done before execution to add guardrails before anything happens. After all steps have been added to the job, they're enumerated over and individually called `Prevalidate()`, as every step will now have fully populated properties to check

Additionally, can be used to ensure that files exist. For reference, see the Prevalidate in BYO cluster creation, ensuring cluster name length is <80chars, as well as ensuring deployment yamls exist in the embedded fs. 

File not found, and naming constraints are classic problems that show up 1hr into execution which can be easily avoided before anything is created

https://github.com/Azure/azure-container-networking/blob/faa2b736341e0ff0366ddfae5f9608cafca29a9e/test/e2e/framework/azure/create-cilium-cluster.go#L45-L69